draft-ietf-core-dns-over-coap.md

```diff
@@ -182,9 +182,14 @@ or automatic configuration, e.g., using a CoRE resource directory
 {{-ddr}}.
 Automatic configuration SHOULD only be done from a trusted source.
 
-Support for SVCB Resource Records {{-svcb}}, {{-svcb-dns}} or DNR Service Parameters {{-dnr}}
-are not specified in this document.
-{{-core-dnr}} explores solutions for CoAP for these mechanisms.
+A DoC server can also be discovered using SVCB Resource Records {{-svcb}}, {{-svcb-dns}} or DNR
+Service Parameters {{-dnr}}.
+\[TBD: draft-lenders-core-coap-dtls-svcb\] provides solutions
+to discover CoAP over (D)TLS servers using the "alpn" SvcParam. This document specifies "docpath" as
+a single-valued SvcParamKey whose value (in both presentation and wire format) MUST be a URI
```

Comment:
I'm wondering whether we can't offer an array of Uri-Path option values in some form.

Reply:
Doing an array with SvcParams would, to my understanding, mean that one needs to provide the "docpath" SvcParamKey multiple times (at least if we don't want to fiddle around with including yet even more data types here). This raises several questions to me: What to do in an absence of a Uri-Path option (i.e. root path)? Not having a docpath key provided? Having it empty? Not providing it at all? Would, for the latter, a client be able to distinguish SvcParams for a DoC server from any other CoAP server?

Reply:
We've discussed briefly whether there would be CRIs or URIs in here; going with URIs removed the ambiguity of earlier versions, but is indeed possibly not ideal.

If multiple values are a thing in SvcParamKey, that would neatly make the default case very efficient, and only become a "burden" if multi-segment paths are used.

The natural thing from a CoAP PoV would be to not provide any at all, bringing us back to the "can we just not set any" discussion we had in a previous iteration, and maybe shied away from too fast: After all, this does not occur in an arbitrary service parameters set (I think it'd make little sense in a hypothetical `_coap SVCB` record, such a record would be used by discovering .well-known/core and then maybe finding a DNS server there), but it is found in a response to a `_dns` SVCB request or the equivalent response from one of the DNRs -- and there, we infer that DoC is present not from the docpath, but from the presence of a CoAP transport in DNS, and then docpath is treated as a plain list, which may be empty.

Reply:
> Doing an array with SvcParams would, to my understanding, mean that one needs to provide the "docpath" SvcParamKey multiple times

Sorry, must have misremembered this from a different spec. Reading RFC 9460 again, it specifically states that duplicate SvcParams are not allowed in a SVCB record:

>   Clients MUST consider an RR malformed if:
>
>   *  the end of the RDATA occurs within a SvcParam.
>
>   *  SvcParamKeys are not in strictly increasing numeric order.
>
>   *  the SvcParamValue for a SvcParamKey does not have the expected
>      format.
>
>   Note that the second condition implies that there are no duplicate
>   SvcParamKeys.

Reply:
So this raises the question: How would an array of URI-Path options be provided?

Reply:
> After all, this does not occur in an arbitrary service parameters set (I think it'd make little sense in a hypothetical `_coap SVCB` record, such a record would be used by discovering .well-known/core and then maybe finding a DNS server there), but it is found in a response to a `_dns` SVCB request or the equivalent response from one of the DNRs -- and there, we infer that DoC is present not from the docpath, but from the presence of a CoAP transport in DNS, and then docpath is treated as a plain list, which may be empty.

I'm still not convinced that an operator might just be allowed to register e.g.

    coap.example.com.  7200  IN SVCB 3 coap.example.com. (
        alpn="co" port="5684" docpath="/" )
       
with no intention to serve anything else but DoC at this endpoint. So the context clues that this is a DoC server would be missing if the SVCB record alone is looked up (of course, they then can use `.well-known/core` to look it up, but that's just one query more). Likewise, it gives an easy hint where to find the DoC server in case any other services are provided at the `coaps://coap.example.com` endpoint.

Reply:
@chrysn and I discussed this and we decided to encode this as a sequence of CBOR text strings. See https://github.com/core-wg/draft-dns-over-coap/pull/28/commits/e08242cf4af5c2a6a29bbb7ccd8936f8a65d5093. The reasoning can also be found there.